File: solutions/ruby/robot-simulator/1/robot_simulator.rb
class Robot

  attr_accessor :bearing
  attr_reader :coordinates
  
  def initialize  # arbitrary initial values?
    @x = 0
    @y = 0
    self.bearing = :north
  end

  MOVES = {west: [-1, 0], north: [0, 1], east: [1, 0], south: [0, -1]}
  DIRECTIONS = MOVES.keys.to_a

  def orient(direction)
    raise ArgumentError unless DIRECTIONS.include?(direction)
    self.bearing = direction
  end

  def turn_right
    self.bearing = DIRECTIONS[(DIRECTIONS.index(self.bearing) + 1) % 4]
  end
  
  def turn_left
    self.bearing = DIRECTIONS[(DIRECTIONS.index(self.bearing) - 1) % 4]
  end

  def at(x, y)
    @x = x
    @y = y
  end

  def advance
    @x += MOVES[self.bearing][0]
    @y += MOVES[self.bearing][1]
  end

  def coordinates
    [@x, @y]
  end
end

class Simulator
  INTERPRET = {'L' => :turn_left, 'R' => :turn_right, 'A' => :advance}
  
  def initialize
  end

  def instructions(sequence)
    sequence.chars.map{|instruction_code| INTERPRET[instruction_code]}
  end

  def place(robot, x:, y:, direction:)
    robot.at(x, y)
    robot.orient(direction)
  end

  def evaluate(robot, instruction_sequence)
    instructions(instruction_sequence).each { |instruction| robot.public_send(instruction) }
  end
  
end

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator {

Robot::Robot(std::pair<int, int> position, Bearing bearing) :
    m_position{position}, m_bearing{bearing} {}

Bearing Robot::get_bearing() const {
    return m_bearing;
}

std::pair<int, int> Robot::get_position() const {
    return m_position;
}

void Robot::turn_right() {
    m_bearing = (Bearing) (((int) m_bearing + 1) % 4);
}

void Robot::turn_left() {
    m_bearing = (Bearing) (((int) m_bearing + 3) % 4);
}

void Robot::advance() {
    switch (m_bearing) {
        case Bearing::NORTH:
            m_position.second++;
            break;
        case Bearing::SOUTH:
            m_position.second--;
            break;
        case Bearing::EAST:
            m_position.first++;
            break;
        case Bearing::WEST:
            m_position.first--;
    };
}

void Robot::execute_sequence(std::string commands) {
    int len = commands.length();
    for (auto idx = 0; idx < len; idx++) {
        switch (commands[idx]) {
            case 'R':
                turn_right();
                break;
            case 'L':
                turn_left();
                break;
            case 'A':
                advance();
        };
    }
}
}  // namespace robot_simulator
